### PR TITLE
Followup fix to transparent inline conversion

### LINKED ISSUE #18123: Regression in implicit resoultion for arguments passed by name
Regression found in Open Communtiy Build in 2 projects:
- stryker-mutator/weapon-regex - [build logs](https://github.com/VirtusLab/community-build3/actions/runs/5426571795/jobs/9870242097) - reproduced below
- typelevel/kittens	- [build logs](https://github.com/VirtusLab/community-build3/actions/runs/5426571795/jobs/9870054515)

## Compiler version
3.3.2-RC1-bin-20230622-9f15be2-NIGHTLY

Bisect points to eb38e1f0fc38313ab17969a74664618900ace2e9

## Minimized code

```Scala

trait P[+T]

extension [T](inline parse0: P[T]) {
  inline def |[V >: T](inline other: P[V]): P[V] = ???
}

// Works if: 
// - parse0 is passed by value or
// - when rep has empty argument list, i.e:  def rep[V]()(using repeater: Implicits.Repeater[T, V]): P[V]
extension [T](inline parse0: => P[T]) {
  inline def rep[V](inline min: Int = 0)(using repeater: Implicits.Repeater[T, V]): P[V] = ???
}

object Implicits {
  trait Repeater[-T, R]
  object Repeater {
    implicit def GenericRepeaterImplicit[T]: Repeater[T, Seq[T]] = ???
  }
}

sealed trait RegexTree
abstract class Node extends RegexTree
class CharClassIntersection() extends Node

def classItem: P[RegexTree] = ???
def charClassIntersection: P[CharClassIntersection] = ???

def x = {
  (charClassIntersection.rep() | classItem.rep())
}
```

## Output

```scala
-- [E172] Type Error: /Users/wmazur/projects/dotty/bisect/main.scala:31:33 -----
31 |  (charClassIntersection.rep() | classItem.rep())
   |                                 ^^^^^^^^^^^^^^^
   |No given instance of type Implicits.Repeater[RegexTree, V] was found.
   |I found:
   |
   |    Implicits.Repeater.GenericRepeaterImplicit[T]
   |
   |But method GenericRepeaterImplicit in object Repeater does not match type Implicits.Repeater[RegexTree, V]
   |
   |where:    V is a type variable with constraint <: Seq[CharClassIntersection]
```

## Expectation
Should compile

### PR DESCRIPTION
**Pull request by julienrf**
Fixes #18123
Follow-up of #17924

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 1]
```diff
@@ -106,16 +106,18 @@ object ProtoTypes {
...
      *  achieved by replacing expected type parameters with wildcards.
      */
     def constrainResult(meth: Symbol, mt: Type, pt: Type)(using Context): Boolean =
-      if (Inlines.isInlineable(meth) && meth.is(Transparent)) {
-        constrainResult(mt, wildApprox(pt))
-        true
+      if (Inlines.isInlineable(meth)) {
+        val methodMatchedType = constrainResult(mt, wildApprox(pt))
+        // do not constrain the result type of transparent inline methods
+        meth.is(Transparent) || methodMatchedType
```
Comment (julienrf) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:120]: I wrote this fix to get a green CI, however I am not convinced this is correct. Why do we have to use `wildApprox(pt)` in the case of non-transparent `inline` methods?

Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:120]: I don't know. It seems non-sensical.

Comment (julienrf) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:120]: So, probably the problem is somewhere else, and we should not change this part?

**Comment by odersky** [comment]
It's a super weird program.

The problem is the `rep` extension method. If that method is `transparent inline`, it works. If it is just inline, it worked before #17924 but does not work now. If it is not inline it does not work. The error is in each case the same. 

So I think these programs really worked by accident. 

If more type variables are instantiated that can be good or bad for implicit search

 - good: fewer ambiguities
 - bad: some cases are thrown out that could have passed if we had let type type variable open.

So, there is no clear answer whether we should constrain more or not. It depends on the circumstances. 

But the point is, https://github.com/lampepfl/dotty/pull/17924 was the correct fix relative to the rationale given in the comment of `constrainResult`. But it seems there's software out that exploited the gap. The fix in the PR is maybe the best one can do. Or else go back to kittens and the other program and tell them to use `transparent inline` for some of their methods.

**Review by odersky — APPROVED** [review]

**Comment by julienrf** [comment]
> Or else go back to kittens and the other program and tell them to use `transparent inline` for some of their methods.

I am considering this option.

What is the impact of making `transparent` an inline method that takes given parameters of type `NotGiven[Something]`? Is this going to cancel the `NotGiven` guarantee (because `wildApprox` is used)?

**Comment by julienrf** [comment]
I’ve submitted the following "fixes" to the affected projects:

https://github.com/stryker-mutator/weapon-regex/pull/230
https://github.com/typelevel/kittens/pull/582

I would rather see those projects fixed instead of having nonsensical code in the compiler. Does anyone has a different opinion?

**Comment by julienrf** [comment]
I took a different approach. Since there is no evidence we should revert what was done in #17924, I've adjusted the fix to still resolve the original issue #9685 but without causing the source incompatibility reported in #18123. The behavior originally implemented in #17924 is still present under `-source future`.

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 2]
```diff
@@ -113,9 +114,22 @@ object ProtoTypes {
      *  achieved by replacing expected type parameters with wildcards.
      */
     def constrainResult(meth: Symbol, mt: Type, pt: Type)(using Context): Boolean =
-      if (Inlines.isInlineable(meth) && meth.is(Transparent)) {
-        constrainResult(mt, wildApprox(pt))
-        true
+      if (Inlines.isInlineable(meth)) {
+        // Stricter behaviour in 3.4+: do not apply `wildApprox` to non-transparent inlines
+        if (Feature.sourceVersion.isAtLeast(SourceVersion.future)) {
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:119]: I guess there would be another follow up to add the `3.4` source version and enable it there?

**Review by bishabosha — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 2 above)
Comment (julienrf) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:119]: Yes, I’ve created #18174 to track that.

**Pull request merged**